### PR TITLE
fix: move uiSchema out of defaults

### PR DESCRIPTION
**Pull request by gjj**
This pull request will attempt to fix https://github.com/TradeTrust/document-creator-website/issues/61:
- changed our schema by having an optional `uiSchema` object on the same level as `schema`
- updated test input

Moving out `uiSchema` from `FormData` to `FormTemplate` instead

#### Unit tests

I've added two unit tests:
1. dynamicForm ⟶ for Covering Letter (DBS) ⟶ should render the remarks field as a textarea
2. dynamicForm ⟶ for Covering Letter (DBS, Nested UISchema) ⟶ should **not** render the remarks field as a textarea

For point 1, `uiSchema` is in the [right level](https://github.com/TradeTrust/document-creator-website/pull/62/files#diff-2b47b5e9ee14c6cf18905418d65f8e13R571) (i.e. same as `defaults` and `formData`) so the Remarks field should be rendered as a `textarea` instead of the default `input`.

For point 2, `uiSchema` is in [`defaults`](https://github.com/TradeTrust/document-creator-website/pull/62/files#diff-2b47b5e9ee14c6cf18905418d65f8e13R571) (as is the case currently and I added on another nested `uiSchema` as requested), which means it's treated as part of `formData` and will show up in the final TradeTrust document output. In this case, `uiSchema` has no effect on the frontend UI.

#### Integration tests
Similarly, I've also added one more integration test that covers the cases above.

In the same test, it will create two similar forms. The first is Covering Letter (DBS):
- Create a new Covering Letter (DBS) form that uses the `form` template's root level `uiSchema` to render the Remarks field as a `textarea`
- Checks if the Remarks field is a `textarea`
- Downloads the file
- Makes sure the downloaded file does not contain `uiSchema` at the root level

Next up, for Covering Letter (DBS, Nested UISchema):
- Create a new Covering Letter (DBS, Nested UISchema) that doesn't contain the `uiSchema` at the root level. Instead, it has `uiSchema` nested in `defaults` and nested again in `misc`
- Checks if the Remarks field is an `input` (and NOT `textarea`)
- Downloads the file
- Makes sure the downloaded file contains two `uiSchema`, one at the root level, and another nested within `misc`

### CODE REVIEW

File: src/types/index.d.ts [hunk 1]
```diff
@@ -45,7 +46,6 @@ export interface FileUploadType {
 // FormData is used by json-schema-forms internally to track state of a single form
 export interface FormData {
   schema?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
-  uiSchema?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (gjj) [inline src/types/index.d.ts:48]: I moved this to `FormTemplate` if that's fine

File: src/components/DynamicFormContainer/DynamicForm/DynamicForm.tsx [hunk 2]
```diff
@@ -70,7 +72,7 @@ export const DynamicFormRaw: FunctionComponent<DynamicFormProps> = ({
     setAttachments(nextAttachment);
   };
 
-  const uiSchema = data.formData.uiSchema || {};
+  uiSchema = uiSchema || {};
```
Comment (yehjxraymond) [inline src/components/DynamicFormContainer/DynamicForm/DynamicForm.tsx:75]: Dont do param reassignment. This may result in unintended effects. Can simply do the defaults in the input itself.

https://stackoverflow.com/questions/23314806/setting-default-value-for-typescript-object-passed-as-argument

While it does not pose a problem by affecting the param passed in directly, if we make this of a habit then there could be bad things that can happen like:
image

File: src/test/fixtures/sample-config.json [hunk 3]
```diff
@@ -573,6 +568,11 @@
           }
         }
       },
+      "uiSchema": {
+        "remarks": {
+          "ui:widget": "textarea"
+        }
+      },
```
Comment (yehjxraymond) [inline src/test/fixtures/sample-config.json:575]: Could you add a case where it's one more level deeper? Just wanted to be certain it works for nested key-values pairs

File: src/types/index.d.ts (hunk 1 above)
Comment (yehjxraymond) [reply src/types/index.d.ts:48]: actually, this is used by json-schema-forms internally right? So the object is actually there during runtime...

Comment (yehjxraymond) [reply src/types/index.d.ts:48]: if thats the case, probably should just leave it there, so at least we know there is uiSchema there, but don't need to use it i guess.

Comment (gjj) [reply src/types/index.d.ts:48]: image

Yeah I did check the docs - it's used internally, but not by `formData`. Rather, it sits on the same level as `formData` in the `Form` props (see [this](https://react-jsonschema-form.readthedocs.io/en/latest/api-reference/form-props/#uischema) and also a visual representation above showing `uiSchema` and `formData` are actually separate from each other).

So it's a bit weird to have `uiSchema` nested within `formData`, but I'm ok to leave it there too.

File: src/components/DynamicFormContainer/DynamicForm/DynamicForm.tsx (hunk 2 above)
Comment (gjj) [reply src/components/DynamicFormContainer/DynamicForm/DynamicForm.tsx:75]: Oh wow that's interesting and TIL, thanks! Will fix

**Comment by gjj** [comment]
I added a new integration test just for `uiSchema`.

It will:
- Create a new Covering Letter (DBS) form that uses the `form` template's root level `uiSchema` to render the Remarks field as a `textarea`
- Checks if the Remarks field is a `textarea`
- Downloads the file
- Makes sure the downloaded file does not contain `uiSchema` at the root level (still figuring out how to compare JSON)

Next up,
- Create a new Covering Letter (DBS, Nested UISchema) that doesn't contain the `uiSchema` at the root level. Instead, it has `uiSchema` nested in `defaults` and nested again in `misc`
- Checks if the Remarks field is an `input` (and NOT `textarea`)
- Downloads the file
- Makes sure the downloaded file contains two `uiSchema`, one at the root level, and another nested within `misc` (still figuring out how to compare JSON)

File: integration/integration-uiSchema.spec.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  // Check that the second Covering Letter has been downloaded successfully
+  await t.click(downloadLink.nth(1));
+  const filePath2 = `${downloadsFolder()}/${fileName2}`;
+  await t.expect(await waitForFileDownload(t, filePath2)).eql(true);
+
+  // We expect the contents of the second doc to contain the uiSchema key,
+  // one at the root level, the other nested in `misc`
+  const docContents2 = readFileSync(filePath2, "utf8");
+  console.log(docContents2);
+  await t.expect(docContents2).contains("uiSchema");
```
Comment (gjj) [inline integration/integration-uiSchema.spec.ts:125]: I couldn't figure out how to use Testcafe to search for specific JSON nodes.

My idea was to expect `docContents2` to contain two `uiSchema` keys. I wanted to check only the keys, so I can't really use `.contains( { "uiSchema": { "remarks": { "ui:widget": "RANDOM_HASH:string:textarea" } } })` because the `RANDOM_HASH` is non-deterministic.

Instead, I am just checking for the presence of only one `uiSchema`, `remarks`, etc. as the next best alternative.

Comment (gjj) [reply integration/integration-uiSchema.spec.ts:125]: So instead of parsing `docContents2` as JSON, I am just leaving it as-is, so that `contains` can just search through a text version of the TradeTrust document.

For reference,
`{"version":"https://schema.openattestation.com/2.0/schema.json","data":{"$template":{"type":"b9423e56-d747-4683-a9ba-7e182bbd1b8c:string:EMBEDDED_RENDERER","name":"5f1de081-7364-4c40-8d33-299fccba3486:string:COVERING_LETTER","url":"c1eaa840-36da-49fe-8065-565dec493d21:string:https://generic-templates.tradetrust.io"},"issuers":[{"name":"a416d63b-313b-4ad9-9d0f-4095b9be04c7:string:Demo Issuer","documentStore":"b1ae58d5-8862-4f95-8516-83da21512f3a:string:0x63a223e025256790e88778a01f480eba77731d04","identityProof":{"type":"73a2aeb1-b8b1-458a-b5fd-90129bd52e2c:string:DNS-TXT","location":"9bbdaa0f-5965-4c7b-8e05-df46590c9d0e:string:demo-tradetrust.openattestation.com"}}],"name":"fa9f5a51-5e58-4b01-a29b-a807610d0b0e:string:Covering Letter","logo":"dee7fbeb-30b2-47d6-bc01-029c228d3410:string:https://pluspng.com/img-png/logo-dbs-png-file-dbs-bank-logo-svg-1280.png","backgroundColor":"67930a3f-fb7d-41f0-947f-3793aef1c149:string:black","titleColor":"63f488e0-2dec-42bf-a9eb-55e4ffd83b22:string:white","remarksColor":"ab9ee80c-4539-4cf0-8e9b-409ad17f1a20:string:white","title":"c8617874-2585-49bc-88d8-e35bb5506a1d:string:Documents Bundle","remarks":"1140c5ab-633e-43e0-9101-3f56cd6d8e7f:string:Some very important documents in here for some submission","misc":{"miscKey1":"ee320769-5843-4411-bcb8-9a83f170fb95:string:miscValue1","uiSchema":{"remarks":{"ui:widget":"4a7fa1d1-a096-4da6-8957-a8f65809bf06:string:textarea"}}},"uiSchema":{"remarks":{"ui:widget":"c003dbdc-2268-4518-a41c-a4a4e56fa6f8:string:textarea"}}},"signature":{"type":"SHA3MerkleProof","targetHash":"3f70323bd6a8bf9f8292e07c01f719156cbc7150c079beb018f524cc67eb5290","proof":["1df26567040c975ccd359137e362cb97e46f6a598d136dc2fc767b3e25d9af84"],"merkleRoot":"f44c0c7ef43f1e947ec159c34ff5d11c1c0804f140bdfb9b90ebb8ed8d420629"}}`

File: scripts/integration-single.sh [hunk 5]
```diff
@@ -1,3 +1,8 @@
+# Install xdg-user-dirs for integration tests  
+if [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    sudo apt-get install -y xdg-user-dirs
```
Comment (gjj) [inline scripts/integration-single.sh:3]: Added this from our integration tests at https://github.com/OpenCerts/opencerts-website/pull/513

We're using `xdg-user-dirs` to get the downloads folder, so that we can have better integration test coverage:

- Download the file
- Check if the file contents are what we're expecting

File: src/test/fixtures/sample-local-config.json [hunk 6]
```diff
@@ -536,6 +536,112 @@
         "allow": true,
         "accept": ".pdf, .json"
       }
+    },
+    {
+      "name": "Covering Letter (DBS)",
```
Comment (gjj) [inline src/test/fixtures/sample-local-config.json:541]: I duplicated this from `sample-config.json` to this file because this `sample-local-config.json` (note the local) is used for our integration tests.

File: integration/integration-end-to-end.spec.ts [hunk 7]
```diff
@@ -133,7 +150,14 @@ test("Upload configuration file, choose form, fill form, submit form correctly",
...
   await t.expect(Title.textContent).contains("Document(s) issued successfully");
-  await t.expect(Selector("div").withText("Document-1-local.tt").exists).ok();
+  await t.expect(Selector("div").withText(fileName).exists).ok();
   await t.expect(Selector("div").withText("Download").exists).ok();
+
+  // Check that the file has been downloaded successfully
+  await t.click(downloadLink);
+  const filePath = `${downloadsFolder()}/${fileName}`;
+  t.ctx.filePath = filePath; // For use in cleanup
+  await t.expect(await waitForFileDownload(t, filePath)).eql(true);
```
Comment (gjj) [inline integration/integration-end-to-end.spec.ts:162]: Improving test case coverage by actually downloading the file, and checking if it's successfully downloaded.

A better case would be to then check the file contents. However, it's tricky because the values are non-deterministic because of the hash.

We can leave that for another PR, if ever needed.

File: src/components/DynamicFormContainer/DynamicForm/DynamicForm.tsx (hunk 2 above)
Comment (gjj) [reply src/components/DynamicFormContainer/DynamicForm/DynamicForm.tsx]: fixed!

File: src/types/index.d.ts (hunk 1 above)
Comment (gjj) [reply src/types/index.d.ts]: added it back!

File: src/test/fixtures/sample-config.json (hunk 3 above)
Comment (gjj) [reply src/test/fixtures/sample-config.json:575]: done, created two cases:
- one where `uiSchema` is at the form template's root
- another where `uiSchema` is nested within `defaults` and within `defaults.misc`

File: src/components/DynamicFormContainer/DynamicForm/DynamicForm.tsx [hunk 8]
```diff
@@ -27,6 +28,7 @@ export interface DynamicFormProps {
 
 export const DynamicFormRaw: FunctionComponent<DynamicFormProps> = ({
   schema,
+  uiSchema = {},
```
Comment (yehjxraymond) [inline src/components/DynamicFormContainer/DynamicForm/DynamicForm.tsx:31]: Is there a need to assign a default to this? Asking because it could alter the behavior of the component if it expects a pass-through.

File: integration/integration-end-to-end.spec.ts (hunk 7 above)
Comment (yehjxraymond) [reply integration/integration-end-to-end.spec.ts:162]: This will be quite similar to just checking if there is a `<a>` with href with data in it right?

File: integration/integration-uiSchema.spec.ts (hunk 4 above)
Comment (yehjxraymond) [reply integration/integration-uiSchema.spec.ts:123]: image

File: integration/integration-end-to-end.spec.ts (hunk 7 above)
Comment (yehjxraymond) [reply integration/integration-end-to-end.spec.ts:162]: Can use JS to check the data file, can run a getData on it so that it's deterministic, probably can skip the oa-verify part though.

File: src/components/DynamicFormContainer/DynamicForm/DynamicForm.test.tsx [hunk 9]
```diff
@@ -4,14 +4,48 @@ import sampleConfig from "../../../test/fixtures/sample-config.json";
...
 import { DynamicForm } from "./DynamicForm";
 
-const form = sampleConfig.forms[0] as FormTemplate;
-
 const mockSetFormData = jest.fn();
 const mockSetOwnership = jest.fn();
 
-const commonProps = {
+// COO
+const form0 = sampleConfig.forms[0] as FormTemplate;
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/components/DynamicFormContainer/DynamicForm/DynamicForm.test.tsx:11]: same as below variables

File: src/test/fixtures/sample-config.json [hunk 10]
```diff
@@ -603,6 +652,14 @@
...
         "title": "Documents Bundle",
         "remarks": "Some very important documents in here for some submission",
+        "misc": {
+          "miscKey1": "miscValue1",
+          "uiSchema": {
+            "remarks": {
+              "ui:widget": "textarea"
+            }
+          }
+        },
```
Comment (yehjxraymond) [inline src/test/fixtures/sample-config.json:662]: image


Hmm what's the impact of this? From the forms perspective I don't see an additional textarea.

Comment (yehjxraymond) [reply src/test/fixtures/sample-config.json:662]: This is to show that if the data has a uiSchema field it wont be affected is it?

File: src/test/fixtures/sample-config.json [hunk 11]
```diff
@@ -603,6 +652,14 @@
...
         "uiSchema": {
           "remarks": {
             "ui:widget": "textarea"
```
Comment (yehjxraymond) [inline src/test/fixtures/sample-config.json:665]: will need to remove right?

File: src/test/fixtures/sample-config.json (hunk 10 above)
Comment (gjj) [reply src/test/fixtures/sample-config.json:662]: Yep - I thought that's what you meant by this: https://github.com/TradeTrust/document-creator-website/pull/62#discussion_r501427631

(correct me if I'm wrong)

File: integration/integration-end-to-end.spec.ts (hunk 7 above)
Comment (gjj) [reply integration/integration-end-to-end.spec.ts:162]: It seems like for file download, the `<a>` with `href` approach is no longer used in favour of `saveAs`. I'll figure how to grab the contents.

File: scripts/integration-single.sh (hunk 5 above)
Comment (gjj) [reply scripts/integration-single.sh:3]: yeah the tricky part is we're not using `href`s anymore

thanks for the SO link! in fact, i even added on our approach for opencerts (using downloads-folder) back in June as there was no other solution

however, it seems like someone from TestCafe added this ~2 weeks ago and created an example in response to the question asked:

> You can inspect how to test file download in TestCafe using [this](https://github.com/DevExpress/testcafe-examples/blob/master/examples/check-downloaded-file-name-and-content/index.js) runnable example. – mlosev Oct 14 at 8:10

Comment (gjj) [reply scripts/integration-single.sh]: Actually - I found a better way. Same approach, but using `os` and `path` to join the Downloads folder.

File: integration/integration-uiSchema.spec.ts (hunk 4 above)
Comment (gjj) [reply integration/integration-uiSchema.spec.ts]: Thank you for this!

File: integration/integration-uiSchema.spec.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+const AddNewButton = Selector("[data-testid='add-new-button']");
+const ProgressBar = Selector("[data-testid='progress-bar']");
+const SubmitButton = Selector("[data-testid='form-submit-button']");
+const DocumentTitleField = Selector("#root_title");
+const DocumentRemarksField = Selector("#root_remarks");
+const goBack = ClientFunction(() => window.history.back());
+const DownloadLink = Selector("[data-testid='download-file-button']");
+
+function getFileDownloadPath(fileName: string): string {
+  return join(homedir(), "Downloads", fileName);
```
Comment (gjj) [inline integration/integration-uiSchema.spec.ts:24]: @yehjxraymond I've since changed to use the `path.join` and `os.homedir` method to get the `Downloads` folder

We can still do a simulated download using this method, except that we don't have to download xdg-user-dirs package
